{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Clade identifier\n",
    "\n",
    "Needs to identify the clades that each strain belongs to in accordance with the naming scheme adopted by\n",
    "Manning 2008. PNAS. Variation in virulence among clades of Escherichia coli O157:H7 associated with disease outbreaks.\n",
    "\n",
    "This program will identify the SNP sites on the different genomes and then turn them into an array. Then, the program will use those to call what genotype they are.\n",
    "\n",
    "First - we "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import Bio as bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline as blastn\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import xml.etree.ElementTree as ET\n",
    "from openpyxl import Workbook\n",
    "from openpyxl.styles import Color, PatternFill, Style\n",
    "from openpyxl import load_workbook\n",
    "\n",
    "#important file locations:\n",
    "sakai = './Clade_data/sakai.fasta'\n",
    "cdata = './Clade_data/'\n",
    "csnps = './Clade_data/SNP_seqs/'\n",
    "cblast = './Clade_data/SNP_BLAST/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#only one genome, can just take it from parser\n",
    "#build a dictionary of locations: 'gene':'location'(string)\n",
    "\n",
    "seqlocs = {}\n",
    "\n",
    "for item in SeqIO.parse(cdata+'sakai.fasta','fasta'):\n",
    "    sakaigenome = item.seq\n",
    "for item in SeqIO.parse(cdata+'sakai.gb','gb'):\n",
    "    for feat in item.features:\n",
    "        feats = dict(feat.qualifiers)\n",
    "        location = str(feat.location)\n",
    "        int1 = int(location.split('[')[1].split(':')[0])\n",
    "        int2 = int(location.split(':')[1].split(']')[0])\n",
    "        if location.split('(')[1] == '-)':\n",
    "            locints = (int2, int1)\n",
    "        else:\n",
    "            locints = (int1, int2)\n",
    "        if ('gene') in feats:\n",
    "            genename = ''.join(feats['gene'])\n",
    "            if not genename in seqlocs:\n",
    "                seqlocs[genename]=locints\n",
    "                \n",
    "print(seqlocs)\n",
    "\n",
    "#checksum - 5498450\n",
    "print('Checksum - 5498450 bases expected: '+str(len(sakaigenome))+' bases found.\\n\\n')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checksum - 5498450 bases expected: 5498450 bases found.\n",
      "\n",
      "\n",
      "Format: Locus Expect/Sakai | []/[]\n",
      "ECs0127 \tC/C\t 1.0\n",
      "ECs0333 \tT/T\t 1.0\n",
      "ECs0472 \tG/C\t 0.6666666666666666\n",
      "ECs0495 \tC/C\t 0.75\n",
      "ECs0517 \tA/A\t 0.8\n",
      "ECs0593 \tT/T\t 0.8333333333333334\n",
      "ECs0606 \tA/T\t 0.7142857142857143\n",
      "ECs0625 \tC/C\t 0.75\n",
      "ECs0654 \tT/A\t 0.6666666666666666\n",
      "ECs0654 \tT/A\t 0.6\n",
      "ECs0655 \tG/C\t 0.5454545454545454\n",
      "ECs0712 \tT/T\t 0.5833333333333334\n",
      "ECs0721 \tG/C\t 0.5384615384615384\n",
      "ECs0915 \tG/C\t 0.5\n",
      "ECs0942 \tA/A\t 0.5333333333333333\n",
      "ECs1028 \tC/C\t 0.5625\n",
      "ECs1262 \tT/A\t 0.5294117647058824\n",
      "ECs1272 \tT/T\t 0.5555555555555556\n",
      "ECs1282 \tC/C\t 0.5789473684210527\n",
      "ECs1496 \tT/T\t 0.6\n",
      "ECs1860 \tG/C\t 0.5714285714285714\n",
      "ECs1895 \tT/T\t 0.5909090909090909\n",
      "ECs2071 \tT/A\t 0.5652173913043478\n",
      "ECs2082 \tG/C\t 0.5416666666666666\n",
      "ECs2357 \tC/C\t 0.56\n",
      "ECs2375 \tC/C\t 0.5769230769230769\n",
      "ECs2521 \tT/T\t 0.5925925925925926\n",
      "ECs2598 \tC/G\t 0.5714285714285714\n",
      "ECs2696 \tC/G\t 0.5517241379310345\n",
      "ECs2775 \tG/G\t 0.5666666666666667\n",
      "ECs2852 \tT/A\t 0.5483870967741935\n",
      "ECs3022 \tT/A\t 0.53125\n",
      "ECs3027 \tC/G\t 0.5151515151515151\n",
      "ECs3044 \tT/A\t 0.5\n",
      "ECs3099 \tA/T\t 0.4857142857142857\n",
      "ECs3221 \tG/C\t 0.4722222222222222\n",
      "ECs3336 \tA/T\t 0.4594594594594595\n",
      "ECs3609 \tC/G\t 0.4473684210526316\n",
      "ECs3635 \tA/A\t 0.46153846153846156\n",
      "ECs3743 \tA/A\t 0.475\n",
      "ECs3788 \tA/A\t 0.4878048780487805\n",
      "ECs3830 \tC/C\t 0.5\n",
      "ECs3880 \tA/T\t 0.4883720930232558\n",
      "ECs3881 \tT/A\t 0.4772727272727273\n",
      "ECs3917 \tT/T\t 0.4888888888888889\n",
      "ECs3942 \tA/A\t 0.5\n",
      "ECs3972 \tG/C\t 0.48936170212765956\n",
      "ECs4022 \tG/G\t 0.5\n",
      "ECs4130 \tT/T\t 0.5102040816326531\n",
      "ECs4251 \tG/G\t 0.52\n",
      "ECs4305 \tA/T\t 0.5098039215686274\n",
      "ECs4380 \tG/C\t 0.5\n",
      "ECs4426 \tA/T\t 0.49056603773584906\n",
      "ECs4479 \tG/G\t 0.5\n",
      "ECs4502 \tC/G\t 0.4909090909090909\n",
      "ECs4589 \tA/T\t 0.48214285714285715\n",
      "ECs4610 \tC/G\t 0.47368421052631576\n",
      "ECs4655 \tA/A\t 0.4827586206896552\n",
      "ECs4667 \tC/G\t 0.4745762711864407\n",
      "ECs4820 \tA/T\t 0.4666666666666667\n",
      "ECs4834 \tA/A\t 0.47540983606557374\n",
      "ECs4889 \tG/G\t 0.4838709677419355\n",
      "ECs5009 \tA/A\t 0.49206349206349204\n",
      "ECs5069 \tA/T\t 0.484375\n",
      "ECs5206 \tG/G\t 0.49230769230769234\n",
      "ECs2006 \tG/G\t 0.5\n",
      "ECs5022 \tA/A\t 0.5074626865671642\n",
      "ECs2082 \tA/T\t 0.5\n",
      "ECs5359 \tT/A\t 0.4927536231884058\n",
      "ECs5359 \tT/A\t 0.4857142857142857\n",
      "ECs1011 \tG/C\t 0.4788732394366197\n",
      "ECs1011 \tG/C\t 0.4722222222222222\n",
      "ECs0492 \tC/C\t 0.4794520547945205\n",
      "ECs4736 \tC/C\t 0.4864864864864865\n",
      "ECs4559 \tC/G\t 0.48\n",
      "ECs4559 \tC/G\t 0.47368421052631576\n",
      "ECs4556 \tT/A\t 0.4675324675324675\n",
      "ECs4556 \tC/G\t 0.46153846153846156\n",
      "ECs2514 \tT/A\t 0.45569620253164556\n",
      "ECs5273 \tT/T\t 0.4625\n",
      "ECs5273 \tC/C\t 0.4691358024691358\n",
      "ECs5273 \tC/C\t 0.47560975609756095\n",
      "ECs5273 \tC/C\t 0.4819277108433735\n",
      "ECs5273 \tC/C\t 0.4880952380952381\n",
      "ECs4109 \tA/T\t 0.4823529411764706\n",
      "ECs4109 \tG/C\t 0.47674418604651164\n",
      "ECs4067 \tC/C\t 0.4827586206896552\n",
      "ECs3595 \tA/T\t 0.4772727272727273\n",
      "ECs3595 \tC/T\t 0.47191011235955055\n",
      "ECs3595 \tA/T\t 0.4666666666666667\n",
      "ECs2325 \tGG/C\t 0.46153846153846156\n",
      "ECs2324 \tC/C\t 0.4673913043478261\n",
      "ECs2325 \tG/G\t 0.4731182795698925\n",
      "ECs5096 \tC/G\t 0.46808510638297873\n",
      "ECs5210 \tA/C\t 0.4631578947368421\n",
      "ECs5298 \tT/A\t 0.4583333333333333\n",
      "44 of 96 ids.\n"
     ]
    }
   ],
   "source": [
    "#DOES NOT WORK - genome does not align. Only around 50% accuracy. Must refine!\n",
    "#Will try to use genbank file to extract gene specific data.\n",
    "\n",
    "#Checksum: extracting releavant locations from sakai genome for blast/muscle\n",
    "#Reads .xlsx file MSU_2096_20SNPs.xlsx, searches the genome, then spits out the two values.\n",
    "#First - load genome data as a large string\n",
    "\n",
    "#only one genome, can just take it from parser\n",
    "for item in SeqIO.parse(cdata+'sakai.fasta','fasta'):\n",
    "    sakaigenome = item.seq\n",
    "\n",
    "#checksum - 5498450\n",
    "print('Checksum - 5498450 bases expected: '+str(len(sakaigenome))+' bases found.\\n\\n')\n",
    "\n",
    "snpxl = load_workbook(filename = cdata+'MSU_2096_20SNPs.xlsx')\n",
    "snpsheet = snpxl['Sheet1']\n",
    "print('Format: Locus Expect/Sakai | []/[]')\n",
    "counter = 0\n",
    "identities = 0\n",
    "for i in range(2,98):\n",
    "    locus = snpsheet['B'+str(i)].value\n",
    "    expect = snpsheet['E'+str(i)].value\n",
    "    sakai = sakaigenome[int(snpsheet['D'+str(i)].value)-1]\n",
    "    counter = counter+1\n",
    "    if expect == sakai:\n",
    "        identities = identities+1\n",
    "    print(locus+' \\t'+expect+'/'+sakai+'\\t '+str(identities/counter))\n",
    "print(str(identities)+' of '+str(counter)+' ids.')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checksum - 5498450 bases expected: 5498450 bases found.\n",
      "\n",
      "\n",
      "Format: Locus Expect/Sakai | []/[]\n",
      "ECs0127\tC/C\t (141454, 143005) \tpositive\t1.0\n",
      "ECs0333\tT/T\t (351026, 351953) \tpositive\t1.0\n",
      "ECs0472\tG/G\t (503108, 501536) \tnegative\t1.0\n",
      "ECs0495\tC/C\t (527966, 529838) \tpositive\t1.0\n",
      "ECs0517\tA/A\t (551825, 552473) \tpositive\t1.0\n",
      "ECs0593\tT/T\t (651425, 652118) \tpositive\t1.0\n",
      "ECs0606\tA/A\t (673459, 672997) \tnegative\t1.0\n",
      "ECs0625\tC/C\t (695812, 699694) \tpositive\t1.0\n",
      "ECs0654\tT/T\t (730925, 729392) \tnegative\t1.0\n",
      "ECs0654\tT/T\t (730925, 729392) \tnegative\t1.0\n",
      "ECs0655\tG/G\t (731859, 730935) \tnegative\t1.0\n",
      "ECs0712\tT/T\t (788855, 789764) \tpositive\t1.0\n",
      "ECs0721\tG/G\t (798793, 796594) \tnegative\t1.0\n",
      "ECs0915\tG/G\t (1003323, 1001997) \tnegative\t1.0\n",
      "ECs0942\tA/A\t (1026908, 1028390) \tpositive\t1.0\n",
      "ECs1028\tC/C\t (1133391, 1133961) \tpositive\t1.0\n",
      "ECs1262\tT/T\t (1323125, 1321781) \tnegative\t1.0\n",
      "ECs1272\tT/T\t (1337360, 1338893) \tpositive\t1.0\n",
      "ECs1282\tC/C\t (1349939, 1353746) \tpositive\t1.0\n",
      "ECs1496\tT/T\t (1536444, 1537689) \tpositive\t1.0\n",
      "ECs1860\tG/G\t (1850890, 1849684) \tnegative\t1.0\n",
      "ECs1895\tT/T\t (1885697, 1887965) \tpositive\t1.0\n",
      "ECs2071\tT/T\t (2061826, 2058085) \tnegative\t1.0\n",
      "ECs2082\tG/G\t (2075111, 2074070) \tnegative\t1.0\n",
      "ECs2357\tC/C\t (2325748, 2326348) \tpositive\t1.0\n",
      "ECs2375\tC/C\t (2346841, 2348446) \tpositive\t1.0\n",
      "ECs2521\tT/T\t (2496633, 2497995) \tpositive\t1.0\n",
      "ECs2598\tC/C\t (2577252, 2575287) \tnegative\t1.0\n",
      "ECs2696\tC/C\t (2653912, 2652994) \tnegative\t1.0\n",
      "ECs2775\tG/G\t (2713478, 2717486) \tpositive\t1.0\n",
      "ECs2852\tT/T\t (2796838, 2795443) \tnegative\t1.0\n",
      "ECs3022\tT/T\t (2954568, 2953956) \tnegative\t1.0\n",
      "ECs3027\tC/C\t (2960670, 2959476) \tnegative\t1.0\n",
      "ECs3044\tT/T\t (2977960, 2976802) \tnegative\t1.0\n",
      "ECs3099\tA/A\t (3039741, 3038094) \tnegative\t1.0\n",
      "ECs3221\tG/G\t (3181451, 3178811) \tnegative\t1.0\n",
      "ECs3336\tA/A\t (3311954, 3309938) \tnegative\t1.0\n",
      "ECs3609\tC/C\t (3599744, 3598820) \tnegative\t1.0\n",
      "ECs3635\tA/A\t (3626014, 3626689) \tpositive\t1.0\n",
      "ECs3743\tA/A\t (3744592, 3745783) \tpositive\t1.0\n",
      "ECs3788\tA/A\t (3800462, 3801458) \tpositive\t1.0\n",
      "ECs3830\tC/C\t (3838186, 3838780) \tpositive\t1.0\n",
      "ECs3880\tA/A\t (3884518, 3883339) \tnegative\t1.0\n",
      "ECs3881\tT/T\t (3885494, 3884507) \tnegative\t1.0\n",
      "ECs3917\tT/T\t (3917392, 3919534) \tpositive\t1.0\n",
      "ECs3942\tA/A\t (3944223, 3944841) \tpositive\t1.0\n",
      "ECs3972\tG/G\t (3981376, 3980824) \tnegative\t1.0\n",
      "ECs4022\tG/G\t (4030353, 4032870) \tpositive\t1.0\n",
      "ECs4130\tT/T\t (4142560, 4144018) \tpositive\t1.0\n",
      "ECs4251\tG/G\t (4252581, 4254903) \tpositive\t1.0\n",
      "ECs4305\tA/A\t (4316045, 4314935) \tnegative\t1.0\n",
      "ECs4380\tG/G\t (4391446, 4389463) \tnegative\t1.0\n",
      "ECs4426\tA/A\t (4453083, 4452552) \tnegative\t1.0\n",
      "ECs4479\tG/G\t (4518483, 4519065) \tpositive\t1.0\n",
      "ECs4502\tC/C\t (4547209, 4546195) \tnegative\t1.0\n",
      "ECs4589\tA/A\t (4620851, 4620725) \tnegative\t1.0\n",
      "ECs4610\tC/C\t (4641014, 4640360) \tnegative\t1.0\n",
      "ECs4655\tA/A\t (4689711, 4690182) \tpositive\t1.0\n",
      "ECs4667\tC/C\t (4703171, 4700636) \tnegative\t1.0\n",
      "ECs4820\tA/A\t (4884458, 4881407) \tnegative\t1.0\n",
      "ECs4834\tA/A\t (4900769, 4901390) \tpositive\t1.0\n",
      "ECs4889\tG/G\t (4963486, 4964860) \tpositive\t1.0\n",
      "ECs5009\tA/A\t (5089179, 5089422) \tpositive\t1.0\n",
      "ECs5069\tA/A\t (5162135, 5161274) \tnegative\t1.0\n",
      "ECs5206\tG/G\t (5306537, 5308040) \tpositive\t1.0\n",
      "ECs2006\tG/G\t (1984118, 1987154) \tpositive\t1.0\n",
      "ECs5022\tA/A\t (5105576, 5106185) \tpositive\t1.0\n",
      "ECs2082\tA/A\t (2075111, 2074070) \tnegative\t1.0\n",
      "ECs5359\tT/T\t (5497104, 5496387) \tnegative\t1.0\n",
      "ECs5359\tT/T\t (5497104, 5496387) \tnegative\t1.0\n",
      "ECs1011\tG/G\t (1115180, 1113989) \tnegative\t1.0\n",
      "ECs1011\tG/G\t (1115180, 1113989) \tnegative\t1.0\n",
      "ECs0492\tC/C\t (523477, 524752) \tpositive\t1.0\n",
      "ECs4736\tC/C\t (4784810, 4787357) \tpositive\t1.0\n",
      "ECs4559\tC/C\t (4599262, 4596457) \tnegative\t1.0\n",
      "ECs4559\tC/C\t (4599262, 4596457) \tnegative\t1.0\n",
      "ECs4556\tT/T\t (4593717, 4593138) \tnegative\t1.0\n",
      "ECs4556\tC/C\t (4593717, 4593138) \tnegative\t1.0\n",
      "ECs2514\tT/T\t (2491575, 2489889) \tnegative\t1.0\n",
      "ECs5273\tT/T\t (5398005, 5398554) \tpositive\t1.0\n",
      "ECs5273\tC/C\t (5398005, 5398554) \tpositive\t1.0\n",
      "ECs5273\tC/C\t (5398005, 5398554) \tpositive\t1.0\n",
      "ECs5273\tC/C\t (5398005, 5398554) \tpositive\t1.0\n",
      "ECs5273\tC/C\t (5398005, 5398554) \tpositive\t1.0\n",
      "ECs4109\tA/A\t (4119505, 4118566) \tnegative\t1.0\n",
      "ECs4109\tG/G\t (4119505, 4118566) \tnegative\t1.0\n",
      "ECs4067\tC/C\t (4077261, 4077540) \tpositive\t1.0\n",
      "ECs3595\tA/A\t (3588074, 3587081) \tnegative\t1.0\n",
      "ECs3595\tC/C\t (3588074, 3587081) \tnegative\t1.0\n",
      "ECs3595\tA/A\t (3588074, 3587081) \tnegative\t1.0\n",
      "ECs2325\tGG/GG\t (2295692, 2294978) \tnegative\t1.0\n",
      "ECs2324\tC/C\t (2294991, 2293878) \tnegative\t1.0\n",
      "ECs2324\tG/G\t (2294991, 2293878) \tnegative\t1.0\n",
      "ECs5096\tC/C\t (5190069, 5188395) \tnegative\t1.0\n",
      "ECs5210\tA/A\t (5311264, 5312638) \tpositive\t1.0\n",
      "ECs5298\tT/T\t (5428996, 5427823) \tnegative\t1.0\n",
      "96 of 96 ids.\n"
     ]
    }
   ],
   "source": [
    "#grab sequence by gene, then find the snp location!\n",
    "#This is a kind of checksum to make sure the right sequences are being pulled at the right index.\n",
    "\n",
    "\n",
    "\n",
    "snpxl = load_workbook(filename = cdata+'MSU_2096_20SNPs.xlsx')\n",
    "snpsheet = snpxl['Sheet1']\n",
    "\n",
    "for item in SeqIO.parse(cdata+'sakai.fasta','fasta'):\n",
    "    sakaigenome = item.seq\n",
    "print('Checksum - 5498450 bases expected: '+str(len(sakaigenome))+' bases found.\\n\\n')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print('Format: Locus Expect/Sakai | []/[]')\n",
    "counter = 0\n",
    "identities = 0\n",
    "for i in range(2,98):\n",
    "    locus = snpsheet['B'+str(i)].value\n",
    "    expect = snpsheet['E'+str(i)].value\n",
    "    geneloc = seqlocs[locus]\n",
    "    genelen = abs(geneloc[0]-geneloc[1])\n",
    "    if int(snpsheet['C'+str(i)].value) < genelen:\n",
    "        if geneloc[0]<geneloc[1]:\n",
    "            geneseq = sakaigenome[geneloc[0]-1:geneloc[1]+1]\n",
    "            posneg = 'positive'\n",
    "        else:\n",
    "            geneseq = sakaigenome[geneloc[1]-1:geneloc[0]+1].reverse_complement()\n",
    "            posneg = 'negative'\n",
    "    else:\n",
    "        print('NOT COOL! NOT COOL, MAN! Skipping '+locus, str(genelen), snpsheet['C'+str(i)].value)\n",
    "        continue\n",
    "    if len(expect) == 1:\n",
    "        sakai = geneseq[snpsheet['C'+str(i)].value]\n",
    "    if len(expect) == 2:\n",
    "        sakai = geneseq[(snpsheet['C'+str(i)].value):(snpsheet['C'+str(i)].value+2)]\n",
    "    counter = counter+1\n",
    "    if expect == sakai:\n",
    "        identities = identities+1\n",
    "    print(locus+'\\t'+expect+'/'+sakai+'\\t',geneloc,'\\t'+posneg+'\\t'+str(identities/counter))\n",
    "print(str(identities)+' of '+str(counter)+' ids.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checksum - 5498450 bases expected: 5498450 bases found.\n",
      "\n",
      "\n",
      "50 1 51 ECs0127 positive C CGAAGGCAATGTCAGTGAAGTGCTGGTGAAGTTTAATCACGACGCACCGAA\n",
      "ECs0127 OKAY!\n",
      "50 1 51 ECs0333 positive T TTTCTGAAATGGGTATTACGCTGGCCGCATTGTCCCAGAATTTAAAATCTC\n",
      "ECs0333 OKAY!\n",
      "50 1 51 ECs0472 negative G GAGCAATTATTATGAGTTCGTCGACTGACGATATTTCAGTGTCATTGAATT\n",
      "ECs0472 OKAY!\n",
      "50 1 51 ECs0495 positive C CCAGATGGGGATGACCGCCGATCAGTACGCCCAGGCGCTGCGTAACCAGCT\n",
      "ECs0495 OKAY!\n",
      "50 1 51 ECs0517 positive A AGCGATCCACTCTCAGTATTAAGAGAGATATTAATTCATGTTCTTGAATCC\n",
      "ECs0517 OKAY!\n",
      "50 1 51 ECs0593 positive T TATCATTACACCGCCACTATTTGTTAGTGAACCCAAAAGCGAAAATACTTT\n",
      "ECs0593 OKAY!\n",
      "50 1 51 ECs0606 negative A ACCCGTTTCACAGGGAACCCTGACACCAGAGCTGTACCAGGAAACGCTGGC\n",
      "ECs0606 OKAY!\n",
      "50 1 51 ECs0625 positive C CGGATGAACACGGTGATTTGAGTATTGAGATCCTCGCCAATAAACAGCATT\n",
      "ECs0625 OKAY!\n",
      "50 1 51 ECs0654 negative T TCGATCGCAAACTGTGCGCCAACCTGGAAGAAGCGATTCGTCGCTCTGGTT\n",
      "ECs0654 OKAY!\n",
      "50 1 51 ECs0654 negative T TCCTGGCGTCCAGCTCCCTGAGTGATTGCCATGCGCCGCTGGTAGAACACA\n",
      "ECs0654 OKAY!\n",
      "50 1 51 ECs0655 negative G GTCCGTGGCATTACGTGAAAAAGACACCGCCCGCCGCATGGTTTATCACGC\n",
      "ECs0655 OKAY!\n",
      "50 1 51 ECs0712 positive T TGATCGCAGCGTCAACGATCTTTATGACGGCACCGCCTGGCTGCAAGCGTT\n",
      "ECs0712 OKAY!\n",
      "50 1 51 ECs0721 negative G GTACGCTTCGAAAAACTGCTGGAATCTGATGCGCCGCTGGCGGAAGTGCTA\n",
      "ECs0721 OKAY!\n",
      "50 1 51 ECs0915 negative G GGCGAGACGGAAGAAGATTTCCAGATGCTACTCGACTTCCTGAAAGAAGCG\n",
      "ECs0915 OKAY!\n",
      "50 1 51 ECs0942 positive A AGAATTATCATAATCGCTACAGTTCGTGGGAGATTGTGCGCGGACAGGAAG\n",
      "ECs0942 OKAY!\n",
      "50 1 51 ECs1028 positive C CGTAAAATCAACCGCCCAACGCGTTGGGCCGGGGGTGAAGCGTCGACAGTG\n",
      "ECs1028 OKAY!\n",
      "50 1 51 ECs1262 negative T TCGCTACTCCCCGGCCCAGCAATACGGAAGAGAAGAAACACACTTTACCAA\n",
      "ECs1262 OKAY!\n",
      "50 1 51 ECs1272 positive T TCGGAAAATGTCCCTGGCAGAAGAAATTAAGAAAGGAATGTCATCCGGAGA\n",
      "ECs1272 OKAY!\n",
      "50 1 51 ECs1282 positive C CAACAACTACGCTACCACCTGGAGTGCGGGCAACCTGGATATCGACGCGAC\n",
      "ECs1282 OKAY!\n",
      "50 1 51 ECs1496 positive T TGATGCGGGTGAAGTGACCAACAGCTATGTTTATATTAAAAGCTGGATTGG\n",
      "ECs1496 OKAY!\n",
      "50 1 51 ECs1860 negative G GATCATTAACCAAATCATTGGTTTTATTGGCTTTCAGGCGCGGACAATTGC\n",
      "ECs1860 OKAY!\n",
      "50 1 51 ECs1895 positive T TACCACGCATCAGGAGGATGAATGAAACTGCAAGGGGTAATTTTCGATCTG\n",
      "ECs1895 OKAY!\n",
      "50 1 51 ECs2071 negative T TGTTAAAAACCTGACATTAGCCGGTGGCAGCACCTGTCCAGTGGTCAGCGT\n",
      "ECs2071 OKAY!\n",
      "50 1 51 ECs2082 negative G GCTGGATGGCATTGAGGTGGTCGGTTCGCTGGTCGGCACGCGCCAGGATCT\n",
      "ECs2082 OKAY!\n",
      "50 1 51 ECs2357 positive C CCGAGCCACTGGAAAACGCGCTGGCACATGTAAAAACCATTATTGCGACGC\n",
      "ECs2357 OKAY!\n",
      "50 1 51 ECs2375 positive C CACTTTCTATTTCTATCTTCGAGCAGGCTGACGAGGCCGGTGTCGGGATGC\n",
      "ECs2375 OKAY!\n",
      "50 1 51 ECs2521 positive T TCCAGCGATCTGCTGCGCGCAGCTTTTCCTGGTGGCTCAATAACCGGGGCT\n",
      "ECs2521 OKAY!\n",
      "50 1 51 ECs2598 negative C CGTGAAGCCGATCTGCATCCACTGGCCGGCGGCGAGCGGGTGCTGGAAGTG\n",
      "ECs2598 OKAY!\n",
      "50 1 51 ECs2696 negative C CGGGGTTATCGTTAAGATTGATGACCTGGGGTATTGGCTGGCGGAAAAATC\n",
      "ECs2696 OKAY!\n",
      "50 1 51 ECs2775 positive G GTTTTGGGTGTTCATACAGCCGAAGCCACACTGCCTAACGGCAACAACGAT\n",
      "ECs2775 OKAY!\n",
      "50 1 51 ECs2852 negative T TGGCGCGCGAGTGAAAAAACTGGTCCATCAACTGGCGGACACCACCTGTTC\n",
      "ECs2852 OKAY!\n",
      "50 1 51 ECs3022 negative T TGCGATGGCACTATCCTGAAGTTATCCTGGTTTACCGGACTGGCGCTGGCT\n",
      "ECs3022 OKAY!\n",
      "50 1 51 ECs3027 negative C CGGGTTTACCACGCCGCAGGGGTAGAACGCCAGGTACGTAATCTGCTGTGG\n",
      "ECs3027 OKAY!\n",
      "50 1 51 ECs3044 negative T TGCCATTCTGGGGATCCTGCTATTAAACATCAGCGCCTTTGGGCTACCAAA\n",
      "ECs3044 OKAY!\n",
      "50 1 51 ECs3099 negative A AGCTGAACGGTGATGTAGCGGCAACAGAACGCGAGTTGCAGTACACCAGCG\n",
      "ECs3099 OKAY!\n",
      "50 1 51 ECs3221 negative G GCACGCTTACCGAAGGGGCCATTGGCTACCGTAAGTTCTCGGTGATCAGCG\n",
      "ECs3221 OKAY!\n",
      "50 1 51 ECs3336 negative A ACAACAGCCGATCCGCTGGGCACAGGGATGCCCGCTGGAAAAAATGGTTAG\n",
      "ECs3336 OKAY!\n",
      "50 1 51 ECs3609 negative C CCACCGCCAGCTCGCCGTTCAGTGGATCAGCTACGGGGAATTGAGGGATCC\n",
      "ECs3609 OKAY!\n",
      "50 1 51 ECs3635 positive A AGTTCTACTGTTAGTCGCCTGGCAGGATCATACCGTGCGTATTGAAATCGG\n",
      "ECs3635 OKAY!\n",
      "50 1 51 ECs3743 positive A ACGTGCTGAAAACATCTCAACCAAAGTCTTTAATAGTGGATTAGGTATTTC\n",
      "ECs3743 OKAY!\n",
      "50 1 51 ECs3788 positive A ATTACCGGCACCCCCGGCGCGGGGAAAAGTACCTTTCTTGAGGCCTTTGGC\n",
      "ECs3788 OKAY!\n",
      "50 1 51 ECs3830 positive C CGTTATTCCGGTGAAGACGCGACCGATCAAAAGAATCTGCAAAAACTGCTG\n",
      "ECs3830 OKAY!\n",
      "50 1 51 ECs3880 negative A ATCTGAACAAGGTGATGTTCTTCATCATCGCGCTCGGTGCGCTGCTGCCGA\n",
      "ECs3880 OKAY!\n",
      "50 1 51 ECs3881 negative T TGGTTGCCGTTACTGCATGGTCGCCTGTCCGTACAACGTGCCGAAGTACGA\n",
      "ECs3881 OKAY!\n",
      "50 1 51 ECs3917 positive T TGCGTAAGCCGTCGTTCGGCAACCAATGGGGCTGACTATTATCTGCCAGGC\n",
      "ECs3917 OKAY!\n",
      "50 1 51 ECs3942 positive A ACCCATTGGCTGGGATCTCACCGGAGTAATGGCGGGAACCTGGTTACTGAC\n",
      "ECs3942 OKAY!\n",
      "50 1 51 ECs3972 negative G GTCGAACTATTGCCGGTTATCGGTACGATTGTCAGTACCTGGGCGCTGTTC\n",
      "ECs3972 OKAY!\n",
      "50 1 51 ECs4022 positive G GCTCTCGCAATATCTGGGCAACGCTTTTGCGCTTATTGATGCTAACGGGGC\n",
      "ECs4022 OKAY!\n",
      "50 1 51 ECs4130 positive T TGGCTTAAGCAACGCAGTACAGACCTTGCAAACCATCGATCCGCAACTGGT\n",
      "ECs4130 OKAY!\n",
      "50 1 51 ECs4251 positive G GCTCCACTTCCCGGACTGGCTGACTATCTTCCTCGCCCAGGGCCTGGGTGG\n",
      "ECs4251 OKAY!\n",
      "50 1 51 ECs4305 negative A AAACCCGGAGCTGACCCAACGCGGTTATCAACACATTATGCGTACTGCCGG\n",
      "ECs4305 OKAY!\n",
      "50 1 51 ECs4380 negative G GGGCTTCTGAAAGCAGCAACCCGATGGTTGATCGTTCAACAATTCAACGCG\n",
      "ECs4380 OKAY!\n",
      "50 1 51 ECs4426 negative A AAATTTTTCCATCACCTATGAGTAAGTTACTGGGCCAGCATCCTGGCCCAA\n",
      "ECs4426 OKAY!\n",
      "50 1 51 ECs4479 positive G GTTATCCATTGACGCCACAGAGAATGGCAGTGTTATTCATATTTCACTGGT\n",
      "ECs4479 OKAY!\n",
      "50 1 51 ECs4502 negative C CTACCTCAATCAAATATCTGGCCAGTATCTATTTATTATCGTTTGCTTTCA\n",
      "ECs4502 OKAY!\n",
      "50 1 51 ECs4589 negative A ACAATACTTCTTAATACGAATGGTACGGTTATGCGGGAAGAGACTAACGCG\n",
      "ECs4589 OKAY!\n",
      "50 1 51 ECs4610 negative C CGGAAAATAATGGTAATCGTACCTGGGTGGCGGCAACTGGAGAGGATGAAT\n",
      "ECs4610 OKAY!\n",
      "50 1 51 ECs4655 positive A AATTTTCTTAATAATGATTTATATAAAAATGATGGAGAGATTCTCTCTTAT\n",
      "ECs4655 OKAY!\n",
      "50 1 51 ECs4667 negative C CACCTGGACGCAATCTTTCACGCCAGGCTGGGGCAATATTTATTTAACATT\n",
      "ECs4667 OKAY!\n",
      "50 1 51 ECs4820 negative A AAAACTACTGGGGCAACTACCCGAAATTCTTCGTCTCTATGATGAAGGCCT\n",
      "ECs4820 OKAY!\n",
      "50 1 51 ECs4834 positive A AGACTCCCCGCTGATGGGTGAAGCTATTTCTGGCGCATCCGGCTTCCCGAT\n",
      "ECs4834 OKAY!\n",
      "50 1 51 ECs4889 positive G GGGCGATTGCTTTTGCAAGGGCGCGGTTGGGGTAAGAACATTTATATGTAC\n",
      "ECs4889 OKAY!\n",
      "50 1 51 ECs5009 positive A AACCACGACCACCAGTACCCAGTAATAAAGTATGTATCCCCAAAATAATTC\n",
      "ECs5009 OKAY!\n",
      "50 1 51 ECs5069 negative A ACATTTTGAATCGGTGGTGCAAGCGTTGCGGCTGGGATTCAGCTCGGTAAT\n",
      "ECs5069 OKAY!\n",
      "50 1 51 ECs5206 positive G GGCGCGGCTGGCTACGTCCCATTTCCCGCAAAGAACAGCAAGAGATTGCCG\n",
      "ECs5206 OKAY!\n",
      "50 1 51 ECs2006 positive G GATGATGGACTCATCCACATCACTAAAGAGGCCGGCGTCGTTGTTGACGGC\n",
      "ECs2006 OKAY!\n",
      "50 1 51 ECs5022 positive A ACGTCACCGTTGTACTAAGAGGAAAAAAATATGGAGTGGAGTCTGACGCAG\n",
      "ECs5022 OKAY!\n",
      "50 1 51 ECs2082 negative A ACAGCATTACCTGTGCGGGGGTCACCACCTACAAAGCCGTTAAGCTGTCAA\n",
      "ECs2082 OKAY!\n",
      "50 1 51 ECs5359 negative T TTTGATCGGCCCTGATGGCGAGCAGTACAAGCTGCCGCGCAGTGAGTTCCG\n",
      "ECs5359 OKAY!\n",
      "50 1 51 ECs5359 negative T TGAGTTCCGCGCCATGCTTCACTTCTGTGAAAACCCAGGCAAAATTCAGTC\n",
      "ECs5359 OKAY!\n",
      "50 1 51 ECs1011 negative G GGTACTGACCAGCGTGAAAAAGGCTGAACAGTATCTGCTCGAAAATGAAAC\n",
      "ECs1011 OKAY!\n",
      "50 1 51 ECs1011 negative G GATCAATGACAAACGTGCTCGCACGGCACAGACTCCGGGTGGCACTGGCGC\n",
      "ECs1011 OKAY!\n",
      "50 1 51 ECs0492 positive C CGGTTCCGGTAAAACGCTGCTGGCTGAAACGCTGGCGCGCCTGCTGGACGT\n",
      "ECs0492 OKAY!\n",
      "50 1 51 ECs4736 positive C CACCCAGCATATACTGCTGCTTGATGAATTTTATCGTACCGCCGTGCGTCT\n",
      "ECs4736 OKAY!\n",
      "50 1 51 ECs4559 negative C CGTTCTGGAGTATCAAGCACTTATAACCTAATAACACAAAACCCTCTTCCT\n",
      "ECs4559 OKAY!\n",
      "50 1 51 ECs4559 negative C CACAAAACCCTCTTCCTGGGGTTAATGTTAATACTCCAAATGTCTATGCGG\n",
      "ECs4559 OKAY!\n",
      "50 1 51 ECs4556 negative T TCCACGTAATGACATAAGTGTAACTGGTATTCGTGATCTTAGTGGTGATTT\n",
      "ECs4556 OKAY!\n",
      "50 1 51 ECs4556 negative C CGTGATCTTAGTGGTGATTTAAGCGCTGGTGATCTGCAAACAGTGAAGGCG\n",
      "ECs4556 OKAY!\n",
      "50 1 51 ECs2514 negative T TCGCCAGGTCAGCCAGGTGAACTTTGCGTCAGAGGTCCACAGGTGATGCTG\n",
      "ECs2514 OKAY!\n",
      "50 1 51 ECs5273 positive T TGATTGATGCGGGTCATACCAACGTTCTGGCTCTGCAGAGTTCAGCTGCGG\n",
      "ECs5273 OKAY!\n",
      "50 1 51 ECs5273 positive C CGCAACAAACGTTGGTGTGCAGATCCTGGACAGAACGGGTGCTGCGCTGAC\n",
      "ECs5273 OKAY!\n",
      "50 1 51 ECs5273 positive C CCAGGCGCGTTATTATGCAATCGGCGAGGCAACCCCGGGTGCTGCTAATGC\n",
      "ECs5273 OKAY!\n",
      "50 1 51 ECs5273 positive C CAGGCGCGTTATTATGCAATCGGCGAGGCAACCCCGGGTGCTGCTAATGCG\n",
      "ECs5273 OKAY!\n",
      "50 1 51 ECs5273 positive C CCTTCAAGGTTCAGTATCAATAACCTACCCAGGTTCAGGGACGTCATTACG\n",
      "ECs5273 OKAY!\n",
      "50 1 51 ECs4109 negative A AGTTGCGAAAACCTGCCCGAAAGCGTGCATTGGTATTATCACTAACCCGGT\n",
      "ECs4109 OKAY!\n",
      "50 1 51 ECs4109 negative G GCACGTTTTGGCCTGTCTCTGGTTCGCGCACTGCAGGGCGAACAAGGCGTT\n",
      "ECs4109 OKAY!\n",
      "50 1 51 ECs4067 positive C CCCCATGAGTTTATCGACAGAACGCAGTTGATGCGTAGCTACACTAAACCG\n",
      "ECs4067 OKAY!\n",
      "50 1 51 ECs3595 negative A AAGCTGGACCATGAACCAAGTGCGGAAGAGATCGCAGAGCAACTGGATAAG\n",
      "ECs3595 OKAY!\n",
      "50 1 51 ECs3595 negative C CATACGCAACCTGGTGGATTCGCCAGACGATTGAACGGGCGATTATGAACC\n",
      "ECs3595 OKAY!\n",
      "50 1 51 ECs3595 negative A AGCACGTGAGTTGTCCCATAAGCTGGACCATGAACCAAGTGCGGAAGAGAT\n",
      "ECs3595 OKAY!\n",
      "50 1 52 ECs2325 negative GG GGGGCACCAGCGGGACTTTGCAAGTGGTGAATCCGCACCTCTGGCAACCGGG\n",
      "ECs2325 OKAY!\n",
      "50 1 51 ECs2324 negative C CCTGAACCGTTATTACGGTTGGTATGTCCAAAGCGGCGATTTGGAAACGGC\n",
      "ECs2324 OKAY!\n",
      "50 1 51 ECs2324 negative G GGCAGGAGAAACTGCATCAGCCGATTATCATCACCGAATACGGCGTGGATA\n",
      "ECs2324 OKAY!\n",
      "50 1 51 ECs5096 negative C CGTCCGACCTATTACAACCGCTATCCGCCGCAGTTCAGGAAATTTACCCCA\n",
      "ECs5096 OKAY!\n",
      "50 1 51 ECs5210 positive A ACTGGGTGCTGGCCGTTGCCGGTACACACGGCAAAACCACCACCGCGGGAA\n",
      "ECs5210 OKAY!\n",
      "50 1 51 ECs5298 negative T TGTTAAAGACGATGCCTCGCGTGCGTTATGGAAAGCCGAGATGCTGCGCTT\n",
      "ECs5298 OKAY!\n"
     ]
    }
   ],
   "source": [
    "#Pull the 50 bases in front of and behind the target SNP.\n",
    "#Some SNPs are too close to end, so must reference whole genome sequence\n",
    "\n",
    "#51st or 51-52nd position of resulting string, (1st or 1st+2nd position of the downstream fragment made)\n",
    "#will have the code for the snp. Has 50 flanking bases BOTH up- and down-stream, so 101b long sequences have snp length 1,\n",
    "#102 have snp length 2, and so on.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "snpxl = load_workbook(filename = cdata+'MSU_2096_20SNPs.xlsx')\n",
    "snpsheet = snpxl['Sheet1']\n",
    "\n",
    "for item in SeqIO.parse(cdata+'sakai.fasta','fasta'):\n",
    "    sakaigenome = item.seq\n",
    "print('Checksum - 5498450 bases expected: '+str(len(sakaigenome))+' bases found.\\n\\n')\n",
    "\n",
    "counter = 0\n",
    "identities = 0\n",
    "for i in range(2,98):\n",
    "    locus = snpsheet['B'+str(i)].value\n",
    "    name = snpsheet['A'+str(i)].value\n",
    "    expect = snpsheet['E'+str(i)].value\n",
    "    geneloc = seqlocs[locus]\n",
    "    genelen = abs(geneloc[0]-geneloc[1])\n",
    "    if int(snpsheet['C'+str(i)].value) < genelen:\n",
    "        if geneloc[0]<geneloc[1]:\n",
    "            snploc = geneloc[0]-1+int(snpsheet['C'+str(i)].value)\n",
    "            posneg = 'positive'\n",
    "        else:\n",
    "            snploc = geneloc[0]-(int(snpsheet['C'+str(i)].value))\n",
    "            posneg = 'negative'\n",
    "    else:\n",
    "        print('NOT COOL! NOT COOL, MAN! Skipping '+locus, str(genelen), snpsheet['C'+str(i)].value)\n",
    "        continue\n",
    "    if (len(expect) == 1) and (posneg=='positive'):\n",
    "        sakaiup = sakaigenome[(snploc-50):(snploc)]\n",
    "        sakaidown = sakaigenome[(snploc):(snploc+51)]\n",
    "    #index error in excel spreadsheet requires the use of a small adjustment here. Also note down/up reversed.\n",
    "    if (len(expect) == 1) and (posneg=='negative'):\n",
    "        sakaidown = sakaigenome[(snploc-50):(snploc+1)].reverse_complement()\n",
    "        sakaiup = sakaigenome[(snploc+1):(snploc+51)].reverse_complement()\n",
    "    if len(expect) == 2:\n",
    "        sakaidown = sakaigenome[(snploc-50):(snploc+2)].reverse_complement()\n",
    "        sakaiup = sakaigenome[(snploc+2):(snploc+52)].reverse_complement()\n",
    "    print(len(sakaiup), len(sakaisnp), len(sakaidown), locus, posneg, expect, sakaidown)\n",
    "    counter = counter+1\n",
    "    if sakaidown[0] == expect:\n",
    "        identities = identities + 1\n",
    "        print(locus, 'OKAY!')\n",
    "    if sakaidown[0:2] == expect:\n",
    "        identities = identities + 1\n",
    "        print(locus, 'OKAY!')\n",
    "    queryseq = SeqRecord(Seq(str(sakaiup+sakaidown), generic_dna))\n",
    "    queryseq.name = name\n",
    "    queryseq.id = name\n",
    "    queryseq.description = locus\n",
    "    queryseqlist = list(queryseq)\n",
    "    handle = open(csnps+name+'.fasta', 'w')\n",
    "    SeqIO.write(queryseq, handle, \"fasta\")\n",
    "    handle.close\n",
    "    \n",
    "    #!!!!!!!!!!!!!!!note - need to run block of code below to get final sequence. No idea why."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 252,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "queryseq = SeqRecord(Seq(str(sakaiup+sakaidown), generic_dna))\n",
    "queryseq.name = name\n",
    "queryseq.id = name\n",
    "queryseq.description = locus\n",
    "queryseqlist = list(queryseq)\n",
    "handle = open(csnps+name+'.fasta', 'w')\n",
    "SeqIO.write(queryseq, handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Need to BLAST all the SNP sequences against the genomes.\n",
    "#d - Database\n",
    "#q - Query\n",
    "#o - output\n",
    "\n",
    "d = './BLAST_DB/DavisDB.fasta'\n",
    "\n",
    "for file in os.listdir(csnps):\n",
    "    if file.endswith('.fasta'):\n",
    "        q =csnps+file\n",
    "        o = cblast+os.path.split(file)[1].split('.')[0]+'.xml'\n",
    "        os.system(str(blastn(query = q, db = d, out = o, outfmt=5)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Extract BLAST hits\n",
    "\n",
    "#hitdict is a dictionary of all blast hits.\n",
    "#it stores a list of important data for each hit for calculation.\n",
    "#hits are checked to be one per contig at end.\n",
    "#hitdict list order:\n",
    "#query start, query stop, query length, alignment length,\n",
    "#number of identical positions, number of 'positive' positions\n",
    "#number of gaps.\n",
    "#[qstart,qstop,qlen,alen,ids,pos,gaps]\n",
    "\n",
    "hitdict = {}\n",
    "\n",
    "#Exclude the salmonella genomes and S54\n",
    "exclude = ('S66', 'S69', 'S70', 'S54')\n",
    "\n",
    "#retrieves xml file, parses with element tree, root = top node\n",
    "for file in os.listdir(cblast):\n",
    "    if file.endswith('.xml'):\n",
    "        querysnp = os.path.splitext(file)[0]\n",
    "        queryf = os.path.join(cblast+file)\n",
    "        tree = ET.parse(queryf)\n",
    "        root = tree.getroot()\n",
    "\n",
    "#Iteration - query information\n",
    "#hit- hits within genome\n",
    "#hsp- alignment infor within a hit\n",
    "        for iteration in root.findall(\".//Iteration\"):\n",
    "            qlen = iteration.find('Iteration_query-len').text\n",
    "        for hit in root.findall(\".//Hit\"):\n",
    "            contig = hit.find('Hit_id').text\n",
    "            if contig.startswith(exclude):\n",
    "                continue\n",
    "            for hsp in hit.findall('.//Hit_hsps/Hsp'):\n",
    "                hseq = hsp.find('Hsp_hseq').text\n",
    "                qstart = hsp.find('Hsp_query-from').text\n",
    "                alen = hsp.find('Hsp_align-len').text\n",
    "                ids = hsp.find('Hsp_identity').text\n",
    "                qseq = hsp.find('Hsp_qseq').text\n",
    "                hseq = hsp.find('Hsp_hseq').text\n",
    "                midl = hsp.find('Hsp_midline').text\n",
    "                \n",
    "            key = contig+'|'+querysnp\n",
    "            if key in list(hitdict.keys()):\n",
    "                print('overwrite at '+key)\n",
    "                \n",
    "            hitdict[key]=[qstart,qlen,alen,ids,qseq,hseq,midl]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING: Genome list does not match hit list for 81_388\n",
      "Missing genome(s): S53\n",
      "SNP has 83 genomes present\n",
      "Unexpected base for 81_388 at S53 Ref/var/test =  A G No hit\n",
      "Unexpected base for 83_1484 at S01 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S02 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S03 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S04 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S05 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S06 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S07 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S08 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S09 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S10 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S11 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S12 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S13 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S14 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S15 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S16 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S17 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S18 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S19 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S20 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S21 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S22 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S23 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S24 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S25 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S26 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S27 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S28 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S29 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S30 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S31 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S32 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S33 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S34 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S37 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S38 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S39 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S40 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S41 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S42 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S43 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S44 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S45 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S46 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S47 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S48 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S49 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S50 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S51 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S52 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S53 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S55 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S56 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S59 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S62 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S63 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S64 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S65 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S67 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S68 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S71 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S72 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S74 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S75 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S76 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S77 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S78 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S81 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S85 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S86 Ref/var/test =  A C G\n",
      "Unexpected base for 83_1484 at S87 Ref/var/test =  A C G\n"
     ]
    }
   ],
   "source": [
    "#Process the data into a spreadsheet\n",
    "#Careful to adjust ends of sequence, then ask if the midline indicates a hit, then store nucleotide\n",
    "#----------------CAUTION----------------CUIDADO----------------------#\n",
    "#Tests need to be recalibrated for new sets of calls!!!!! They were made here not to solve all problems,\n",
    "#but only the problems present in this dataset!!!!!\n",
    "\n",
    "wb = Workbook()\n",
    "ws = wb.create_sheet(0)\n",
    "\n",
    "#Create column and row labels\n",
    "\n",
    "snps = []\n",
    "genomes = ['0','1',]\n",
    "\n",
    "for item in hitdict:\n",
    "    genomes.append(item.split('(')[0])\n",
    "    snps.append(item.split('|')[1])\n",
    "genomes = sorted(set(genomes))\n",
    "snps = sorted(set(snps))\n",
    "genomes[0] = 'Sakai'\n",
    "genomes[1] = 'Variant'\n",
    "\n",
    "for n in range(0, len(snps)):\n",
    "    ws.cell(row = (n+2), column = 1).value = snps[n]\n",
    "    \n",
    "for n in range(0, len(genomes)):\n",
    "    ws.cell(row = 1, column = (n+2)).value = genomes[n]\n",
    "\n",
    "#Add in expected sakai and deviant values\n",
    "snpxl = load_workbook(filename = cdata+'MSU_2096_20SNPs.xlsx')\n",
    "snpsheet = snpxl['Sheet1']\n",
    "for i in range(2,98):\n",
    "    locus = snpsheet['A'+str(i)].value\n",
    "    expect = snpsheet['E'+str(i)].value\n",
    "    variant = snpsheet['F'+str(i)].value\n",
    "    ws.cell(row = snps.index(locus)+2, column=2).value = expect\n",
    "    ws.cell(row = snps.index(locus)+2, column=3).value = variant\n",
    "\n",
    "#hitdictkeys is a dict collected by snp of all of the hits\n",
    "hitdictkeys = {}\n",
    "hitdictkeylist = list(hitdict.keys())\n",
    "for item in hitdictkeylist:\n",
    "    if not item.split('|')[1] in hitdictkeys:\n",
    "        hitdictkeys[item.split('|')[1]]=[]\n",
    "    hitdictkeys[item.split('|')[1]].append(item)\n",
    "\n",
    "for snp in snps:\n",
    "    if not (len(hitdictkeys[snp])==(len(genomes)-2)):\n",
    "        print('WARNING: SNP does not have enough sequences!')\n",
    "    templist = []\n",
    "    for hit in hitdictkeys[snp]:\n",
    "        templist.append(hit.split('(')[0])\n",
    "    templist = sorted(set(templist))\n",
    "    if genomes[2:]!=templist:\n",
    "        print('WARNING: Genome list does not match hit list for',snp)\n",
    "        print('Missing genome(s):',','.join([x for x in genomes[2:] if x not in templist]))\n",
    "        print('SNP has',len(templist),'genomes present')\n",
    "        ws.cell(row = snps.index('81_388')+2, column = genomes.index('S53')+2).value = 'No hit'\n",
    "    \n",
    "    #REMINDER - hitdict[key]=[qstart,qlen,alen,ids,qseq,hseq,midl]\n",
    "    keylist = list(hitdictkeys[snp])\n",
    "    for key in keylist:\n",
    "        #position - the expected position is 50, adjusted for hits\n",
    "        position = 50\n",
    "        \n",
    "        qstart = int(hitdict[key][0])\n",
    "        qlen = int(hitdict[key][1])\n",
    "        alen = int(hitdict[key][2])\n",
    "        ids = int(hitdict[key][3])\n",
    "        qseq = hitdict[key][4]\n",
    "        hseq = hitdict[key][5]\n",
    "        midl = hitdict[key][6]\n",
    "        #locus - snp ID, genome - genome hit is in\n",
    "        locus = key.split('|')[1]\n",
    "        genome = key.split('(')[0]\n",
    "        if (qlen==101) and (alen==101) and ('-' not in (hseq or qseq)):\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = hseq[position]\n",
    "            continue\n",
    "        if (qlen==101) and (qstart!=1) and ((alen+(qstart-1))==101) and (' ' not in midl) and ('-' not in (hseq or qseq)):\n",
    "            position = 50-(qstart-1)\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = hseq[position]\n",
    "            continue\n",
    "        if (qlen==102) and (alen==102) and ('-' not in (hseq or qseq)):\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = ''.join(hseq[50:52])\n",
    "            continue\n",
    "        #Problem with align 26_510 - insertion upstream of SNP (T a position -2 for SNP in two isolates)\n",
    "        #This solves that. overkill selectors, basically needs to ignore insertion to get SNP.\n",
    "        if (qlen==101) and ('-' in qseq) and ('-' not in hseq) and (qseq.index('-')<50) and (qseq.count('-')==1) and (locus=='26_510'):\n",
    "            position = position+1\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = hseq[position]\n",
    "            continue\n",
    "        if ('-' in hseq) and ('-' not in qseq) and (qlen==101):\n",
    "            position = 50-(qstart-1)\n",
    "            if hseq[position] == '-':\n",
    "                position = position+1\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = hseq[position]\n",
    "            continue\n",
    "        #no blast hit for SNP area - 81_388/S24 had to be investigated by hand. Has 2 partial hits, neither with SNP region\n",
    "        if (locus==('81_388') and (genome=='S24')):\n",
    "            ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value = 'No hit'\n",
    "            continue\n",
    "        print('Did not find base for:', key)\n",
    "\n",
    "#Workbook check - see if values match either SNP or Sakai, flag if not\n",
    "#li - locus index\n",
    "#gi - genome index\n",
    "\n",
    "for locus in snps:\n",
    "    for genome in genomes[2:]:\n",
    "        sakai = ws.cell(row = snps.index(locus)+2, column = 2).value\n",
    "        variant = ws.cell(row = snps.index(locus)+2, column = 3).value\n",
    "        test =  ws.cell(row = snps.index(locus)+2, column = genomes.index(genome)+2).value\n",
    "        if (test == sakai) or (test == variant):\n",
    "            continue\n",
    "        print('Unexpected base for', locus, 'at', genome, 'Ref/var/test = ', sakai, variant, test)\n",
    "        #83_1484 SNP that this test returns for this dataset is LEGIT! Woo!\n",
    "\n",
    "\n",
    "##########################################################\n",
    "        \n",
    "#Now need to compare the results from this output to a preexistig matrix of hits, then call what clade they are in.\n",
    "#File SNP_panel_edited.xlsx has a table of SNPs for this purpose.\n",
    "#It has one row deleted compared to SNP_panel.xlsx that was gotten from\n",
    "#Mark Mammel at the FDA in Dec. 2015.\n",
    "\n",
    "#It may be useful to create a crosstable of %IDs\n",
    "\n",
    "#First, we will build a series of strings.\n",
    "#Letters are the respective nucleotides\n",
    "#where an 'N' stands for 'no call' and is ignored in the calculation if either has one.\n",
    "\n",
    "#Using the workbook made in the cell above.\n",
    "\n",
    "#Load new workbook SNP_panel_edited.xlsx\n",
    "\n",
    "snprefwb = load_workbook(filename = cdata+'SNP_panel_edited.xlsx')\n",
    "snppanel = snprefwb['genotypes']\n",
    "#snpids has the SNP names in order on the sheet(best order for identification)\n",
    "#refdict - has strings of nuclotides for each clade reference. IDs are stored as '# clade' with space separator\n",
    "#snpdict - strings for each test genome\n",
    "#cladeids - clade names in order (best order for identification)\n",
    "#snprow - list of snpnames in workbook from cell above - used to find index of each SNP by name\n",
    "snpids = []\n",
    "refdict = {}\n",
    "snpdict = {}\n",
    "cladeids = []\n",
    "snprow = []\n",
    "for i in range(2,60):\n",
    "    snpids.append(snppanel.cell(row=i, column=3).value)\n",
    "for i in range(4,40):\n",
    "    cladeids.append(snppanel.cell(row=1, column=i).value)\n",
    "for i in range(4,40):\n",
    "    claderef = snppanel.cell(row=1, column=i).value\n",
    "    refdict[claderef]=''\n",
    "    for j in range(2,60):\n",
    "        refdict[claderef]=refdict[claderef]+snppanel.cell(row=j, column=i).value\n",
    "\n",
    "for i in range(2,98):\n",
    "    snprow.append(snpsheet.cell(row = i, column=1).value)\n",
    "for i in range(2,88):\n",
    "    genome = ws.cell(row=1, column=i).value\n",
    "    snpdict[genome]=''\n",
    "    for item in snpids:\n",
    "        if ws.cell(row=snprow.index(item)+2, column=i).value[0] == ' ':\n",
    "            snpdict[genome]=snpdict[genome]+'-'\n",
    "        else:\n",
    "            snpdict[genome]=snpdict[genome]+ws.cell(row=snprow.index(item)+2, column=i).value[0]\n",
    "\n",
    "#Build an array of values of %IDs for each genome SNP set vs references\n",
    "#Will use a spreadsheet to store these values. First store each snp in ws2\n",
    "#Store %ID in sheet ws3\n",
    "#Could build a regular matrix first, but nah.\n",
    "\n",
    "ws2 = wb.create_sheet(1)\n",
    "ws3 = wb.create_sheet(2)\n",
    "\n",
    "for item in snpids:\n",
    "    ws2.cell(row = snpids.index(item)+2, column = 1).value = item\n",
    "for item in genomes:\n",
    "    ws2.cell(row = 1, column = genomes.index(item)+2).value=item\n",
    "for item in genomes:    \n",
    "    for jtem in snpids:\n",
    "        ws2.cell(row = snpids.index(jtem)+2, column = genomes.index(item)+2).value = snpdict[item][snpids.index(jtem)]\n",
    "\n",
    "for genome in genomes:\n",
    "    ws3.cell(row = genomes.index(genome)+2, column = 1).value = genome\n",
    "for clade in cladeids:\n",
    "    ws3.cell(row = 1, column = cladeids.index(clade)+2,).value = clade\n",
    "for genome in genomes:\n",
    "    for clade in cladeids:\n",
    "        ids = 0\n",
    "        for i in range(0,58):\n",
    "            if (snpdict[genome][i]) == (refdict[clade][i]):\n",
    "                ids += 1\n",
    "        pctids = 100*ids/58\n",
    "        ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value = pctids\n",
    "fill1= PatternFill(fill_type='solid', start_color='FF9999', end_color='FF9999')\n",
    "#Mark the highest values\n",
    "for genome in genomes:\n",
    "    pcts = []\n",
    "    for clade in cladeids:\n",
    "        pcts.append(float(ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value))\n",
    "    bestvalue = max(pcts)\n",
    "    for clade in cladeids:\n",
    "        if float(ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value) == bestvalue:\n",
    "            ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).fill = fill1\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "        \n",
    "wb.save('SNP_matrix.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Now need to compare the results from this output to a preexistig matrix of hits, then call what clade they are in.\n",
    "#File SNP_panel_edited.xlsx has a table of SNPs for this purpose.\n",
    "#It has one row deleted compared to SNP_panel.xlsx that was gotten from\n",
    "#Mark Mammel at the FDA in Dec. 2015.\n",
    "\n",
    "#It may be useful to create a crosstable of %IDs\n",
    "\n",
    "#First, we will build a series of strings.\n",
    "#Letters are the respective nucleotides\n",
    "#where an 'N' stands for 'no call' and is ignored in the calculation if either has one.\n",
    "\n",
    "#Using the workbook made in the cell above.\n",
    "\n",
    "#Load new workbook SNP_panel_edited.xlsx\n",
    "\n",
    "snprefwb = load_workbook(filename = cdata+'SNP_panel_edited.xlsx')\n",
    "snppanel = snprefwb['genotypes']\n",
    "#snpids has the SNP names in order on the sheet(best order for identification)\n",
    "#refdict - has strings of nuclotides for each clade reference. IDs are stored as '# clade' with space separator\n",
    "#snpdict - strings for each test genome\n",
    "#cladeids - clade names in order (best order for identification)\n",
    "#snprow - list of snpnames in workbook from cell above - used to find index of each SNP by name\n",
    "snpids = []\n",
    "refdict = {}\n",
    "snpdict = {}\n",
    "cladeids = []\n",
    "snprow = []\n",
    "for i in range(2,60):\n",
    "    snpids.append(snppanel.cell(row=i, column=3).value)\n",
    "for i in range(4,40):\n",
    "    cladeids.append(snppanel.cell(row=1, column=i).value)\n",
    "for i in range(4,40):\n",
    "    claderef = snppanel.cell(row=1, column=i).value\n",
    "    refdict[claderef]=''\n",
    "    for j in range(2,60):\n",
    "        refdict[claderef]=refdict[claderef]+snppanel.cell(row=j, column=i).value\n",
    "\n",
    "for i in range(2,98):\n",
    "    snprow.append(snpsheet.cell(row = i, column=1).value)\n",
    "for i in range(2,88):\n",
    "    genome = ws.cell(row=1, column=i).value\n",
    "    snpdict[genome]=''\n",
    "    for item in snpids:\n",
    "        if ws.cell(row=snprow.index(item)+2, column=i).value[0] == ' ':\n",
    "            snpdict[genome]=snpdict[genome]+'-'\n",
    "        else:\n",
    "            snpdict[genome]=snpdict[genome]+ws.cell(row=snprow.index(item)+2, column=i).value[0]\n",
    "\n",
    "#Build an array of values of %IDs for each genome SNP set vs references\n",
    "#Will use a spreadsheet to store these values. First store each snp in ws2\n",
    "#Store %ID in sheet ws3\n",
    "#Could build a regular matrix first, but nah.\n",
    "\n",
    "ws2 = wb.create_sheet(1)\n",
    "ws3 = wb.create_sheet(2)\n",
    "\n",
    "for item in snpids:\n",
    "    ws2.cell(row = snpids.index(item)+2, column = 1).value = item\n",
    "for item in genomes:\n",
    "    ws2.cell(row = 1, column = genomes.index(item)+2).value=item\n",
    "for item in genomes:    \n",
    "    for jtem in snpids:\n",
    "        ws2.cell(row = snpids.index(jtem)+2, column = genomes.index(item)+2).value = snpdict[item][snpids.index(jtem)]\n",
    "\n",
    "for genome in genomes:\n",
    "    ws3.cell(row = genomes.index(genome)+2, column = 1).value = genome\n",
    "for clade in cladeids:\n",
    "    ws3.cell(row = 1, column = cladeids.index(clade)+2,).value = clade\n",
    "for genome in genomes:\n",
    "    for clade in cladeids:\n",
    "        ids = 0\n",
    "        for i in range(0,58):\n",
    "            if (snpdict[genome][i]) == (refdict[clade][i]):\n",
    "                ids += 1\n",
    "        pctids = 100*ids/58\n",
    "        ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value = pctids\n",
    "fill1= PatternFill(fill_type='solid', start_color='FF9999', end_color='FF9999')\n",
    "#Mark the highest values\n",
    "for genome in genomes:\n",
    "    pcts = []\n",
    "    for clade in cladeids:\n",
    "        pcts.append(float(ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value))\n",
    "    bestvalue = max(pcts)\n",
    "    for clade in cladeids:\n",
    "        if float(ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).value) == bestvalue:\n",
    "            ws3.cell(row = genomes.index(genome)+2, column = cladeids.index(clade)+2).fill = fill1\n",
    "        \n",
    "    \n",
    "        \n",
    "        \n",
    "        \n",
    "wb.save('SNP_matrix.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "check = []\n",
    "ids = (99, 100, 101)\n",
    "for item in hitdict:\n",
    "    if '43_3971' in item:\n",
    "        print (hitdict[item])\n",
    "        check.append(item)\n",
    "        if (hitdict[item][3] == ('101' or '100')) and (hitdict[item][4]==('99' or '100' or '101')):\n",
    "            check.append(item)\n",
    "print(len(check))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'S74': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S02': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S10': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S50': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S43': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S22': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S34': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S25': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S23': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S65': 'GCGGGCGCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S45': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S36': 'GCGGGCTCCGGTGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S72': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S15': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S56': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S59': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S81': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S04': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S46': 'GCGGGCGCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S88': 'GCGGGCTCCGGTGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S67': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S80': 'GCGGGCTCCGGTGGGCCGAGGCCGTTGAGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S79': 'GCGGGCTCCGGTGGGCCGAGGCCGTTGAGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S21': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S76': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S73': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S51': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S41': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S05': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S86': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S24': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S17': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'Sakai': 'TTAAATGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S63': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S20': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S77': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S03': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S78': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S35': 'GCGGGCTCCGGTGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S27': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S29': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S68': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S26': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'Variant': 'GCGGGCTCCGGCCAGCCGAGTCCGTTGAAAGACCCCAAATTATATTCAC-TAAATATC', 'S48': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S12': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S28': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S58': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S42': 'GCGGGCTCCGGTGGATGGGAGCGACCAGGGAAATCTTAGGTCCGCCTCTGCGGCCCCT', 'S37': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S64': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S57': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S82': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S53': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S55': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S87': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S14': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S84': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S83': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S09': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S85': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S61': 'GCGGGCGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S39': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S33': 'GCGGGCGCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S49': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S44': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S11': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S08': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S52': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S60': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', 'S30': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S71': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S40': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S06': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S62': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S31': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S38': 'GCGGGCGCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S75': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S13': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S18': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S19': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S47': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', 'S01': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S16': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S07': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', 'S32': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT'}\n",
      "{'26 6': 'GCGGGCTCCGGTGGGCCGAGGCCGTTGAAGACATATTCGGCCCGCCTCTGCGGCCCCT', '23 5': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '36 9.2': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCCACCCAACCCCT', '02 1': 'GTAAATGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '04 2': 'GCGGGTGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '24 6': 'GCGGGCTCCGGTGGGCCGAGGCCGTTGAGGACATATTCGGCCCGCCTCTGCGGCCCCT', '10 2': 'GCGGGCTTTCACGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '28 7': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACTATTCGGCCCGCCTCTGCGGCCCCT', '13 3': 'GCGGGCTCCGGTCGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '35 9.1': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCCACCTGGCCCCT', '17 3': 'GCGGGCTCCGGTGGGTGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '33 8': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATCATATCTCTGCGGCCCCT', '39 9.3': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCCACCCAGATACT', '11 2': 'GCGGGCTTTCACGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '32 8': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATCATACCTCTGCGGCCCCT', '03 2': 'GCGGGTGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '30 8': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATTCCGCCTCTGCGGCCCCT', '34 8': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATCATACTTCTGCGGCCCCT', '19 4': 'GCGGGCTCCGGTGGGCCGAGTTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '12 3': 'GCGGGCTCCGGTGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '16 3': 'GCGGGCTCCGGTGGGTGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '29 7': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCTCTGCGGCCCCT', '31 8': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCCCAAATCACGCCTCTGCGGCCCCT', '07 2': 'GCGGGTGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '05 2': 'GCGGGTGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '37 9.4': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCCACCCAGACCTC', '38 9.3': 'GCGGGCTCCGGTGGGCCGAGGCCGCCAGGGGACCATTCGGCCCGCCCACCCAGATCCT', '18 3': 'GCGGGCTCCGGTGGGTGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '09 2': 'GCGGGCTTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '01 1': 'TTAAATGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '06 2': 'GCGGGTGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '15 3': 'GCGGGCTCCGGTCGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '14 3': 'GCGGGCTCCGGTCAATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '25 6': 'GCGGGCTCCGGTGGGCCGAGGCCGTTGAAAACATATTCGGCCCGCCTCTGCGGCCCCT', '20 4': 'GCGGGCTCCGGTGGGCCGAGGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT', '08 2': 'GCGGGCGTTCATGGATGAGAGTGACCAGGGACATATTCGGCCCGCCTCTGCGGCCCCT'}\n",
      "['Sakai', 'Variant', 'S01', 'S02', 'S03', 'S04', 'S05', 'S06', 'S07', 'S08', 'S09', 'S10', 'S11', 'S12', 'S13', 'S14', 'S15', 'S16', 'S17', 'S18', 'S19', 'S20', 'S21', 'S22', 'S23', 'S24', 'S25', 'S26', 'S27', 'S28', 'S29', 'S30', 'S31', 'S32', 'S33', 'S34', 'S35', 'S36', 'S37', 'S38', 'S39', 'S40', 'S41', 'S42', 'S43', 'S44', 'S45', 'S46', 'S47', 'S48', 'S49', 'S50', 'S51', 'S52', 'S53', 'S55', 'S56', 'S57', 'S58', 'S59', 'S60', 'S61', 'S62', 'S63', 'S64', 'S65', 'S67', 'S68', 'S71', 'S72', 'S73', 'S74', 'S75', 'S76', 'S77', 'S78', 'S79', 'S80', 'S81', 'S82', 'S83', 'S84', 'S85', 'S86', 'S87', 'S88']\n",
      "['01 1', '02 1', '03 2', '04 2', '06 2', '05 2', '07 2', '08 2', '09 2', '10 2', '11 2', '12 3', '13 3', '14 3', '15 3', '16 3', '17 3', '18 3', '19 4', '20 4', '23 5', '24 6', '25 6', '26 6', '28 7', '29 7', '30 8', '31 8', '32 8', '33 8', '34 8', '35 9.1', '36 9.2', '38 9.3', '39 9.3', '37 9.4']\n"
     ]
    }
   ],
   "source": [
    "print(snpdict)\n",
    "print(refdict)\n",
    "print(genomes)\n",
    "print(cladeids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G\n",
      "C\n",
      "G\n",
      "G\n",
      "G\n",
      "C\n",
      "T\n",
      "C\n",
      "C\n",
      "G\n",
      "G\n",
      "T\n",
      "G\n",
      "G\n",
      "G\n",
      "C\n",
      "C\n",
      "G\n",
      "A\n",
      "G\n",
      "G\n",
      "C\n",
      "C\n",
      "G\n",
      "C\n",
      "C\n",
      "A\n",
      "G\n",
      "G\n",
      "G\n",
      "G\n",
      "A\n",
      "C\n",
      "C\n",
      "C\n",
      "C\n",
      "A\n",
      "A\n",
      "A\n",
      "T\n",
      "T\n",
      "C\n",
      "C\n",
      "G\n",
      "C\n",
      "C\n",
      "T\n",
      "C\n",
      "T\n",
      "G\n",
      "C\n",
      "G\n",
      "G\n",
      "C\n",
      "C\n",
      "C\n",
      "C\n",
      "T\n"
     ]
    }
   ],
   "source": [
    "for i in range(len(snpdict['S32'])):\n",
    "    print(snpdict['S32'][i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n"
     ]
    }
   ],
   "source": [
    "for i in range(58):\n",
    "    print (i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
